miniapps/gslib/ex1p.cpp

```diff
@@ -0,0 +1,454 @@
+   // corresponding to the Laplacian operator -Delta, by adding the Diffusion
+   // domain integrator.
+   ParBilinearForm *a = new ParBilinearForm(fespace);
+   a->AddDomainIntegrator(new DiffusionIntegrator(one));
+
+   // Assemble the bilinear form and the corresponding linear system,
+   // applying any necessary transformations such as: eliminating boundary
+   // conditions, applying conforming constraints for non-conforming AMR,
+   // static condensation, etc.
+   a->Assemble();
+
+   delete b;
+
+   // Use simultaneous Schwarz iterations to iteratively solve the PDE
+   // and interpolate interdomain boundary data to impose Dirichlet
+   // boundary conditions.
+
+   int NiterSchwarz = 100;
+   for (int schwarz = 0; schwarz < NiterSchwarz; schwarz++)
+   {
+      ParLinearForm *b = new ParLinearForm(fespace);
+      b->AddDomainIntegrator(new DomainLFIntegrator(one));
+      b->Assemble();
+
+      OperatorPtr A;
+      Vector B, X;
+      a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
+
+      // Solve the linear system A X = B.
+      // Use a simple symmetric Gauss-Seidel preconditioner with PCG.
+      Solver *prec = NULL;
+      prec = new HypreBoomerAMG;
+      CGSolver cg(*comml);
+      cg.SetRelTol(1e-12);
+      cg.SetMaxIter(2000);
+      cg.SetPrintLevel(0);
+      if (prec) { cg.SetPreconditioner(*prec); }
+      cg.SetOperator(*A);
+      cg.Mult(B, X);
+      delete prec;
+
+      // Recover the solution as a finite element grid function.
+      a->RecoverFEMSolution(X, *b, x);
+
+      // Interpolate boundary condition
+      finder.Interpolate(x, interp_vals1);
+
+      double dxmax = std::numeric_limits<float>::min();
+      double xinf = x.Normlinf();
+      double xinfg = xinf;
+      MPI_Allreduce(&xinf, &xinfg, 1, MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);
+      x.SetTrueVector();
+      Vector xt = x.GetTrueVector();
+      for (int i = 0; i < nb1; i++)
+      {
+         int idx = ess_tdof_list_int[i];
+         double dx = std::abs(xt(idx)-interp_vals1(i))/xinfg;
+         if (dx > dxmax) { dxmax = dx; }
+         xt(idx) = interp_vals1(i);
+      }
+      x.SetFromTrueDofs(xt);
+      double dxmaxg = dxmax;
+      MPI_Allreduce(&dxmax, &dxmaxg, 1, MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);
+
+      delete b;
+
+      if (myid == 0)
+      {
+         std::cout << std::setprecision(8)    <<
+                   "Iteration: "           << schwarz <<
+                   ", Relative residual: " << dxmaxg   << endl;
+      }
+
+      if (dxmaxg < rel_tol) { break; }
+   }
+
+   {
+      // output files
+      ostringstream mesh_name, sol_name;
+      mesh_name << "og_mesh." << setfill('0') << setw(6) << myid;
+      sol_name << "og_sol." << setfill('0') << setw(6) << myid;
+
+      ofstream mesh_ofs(mesh_name.str().c_str());
+      mesh_ofs.precision(8);
+      pmesh->Print(mesh_ofs);
+
+      ofstream sol_ofs(sol_name.str().c_str());
+      sol_ofs.precision(8);
+      x.Save(sol_ofs);
```

Comment:
mesh and solutions files need to be added to the `make clean` section in the makefile